src/irr/asset/CGLIWriter.cpp

```diff
@@ -0,0 +1,327 @@
+			const auto channelsAmount = getFormatChannelCount(imageInfo.format);
+			const auto singleChannelByteSize = texelBlockByteSize / channelsAmount;
+			const auto data = reinterpret_cast<const uint8_t*>(image->getBuffer()->getPointer());
+
+			auto getTarget = [&]()
+			{
+				switch (imageViewInfo.viewType)
+				{
+					case ICPUImageView::ET_1D: return gli::TARGET_1D;
+					case ICPUImageView::ET_1D_ARRAY: return gli::TARGET_1D_ARRAY;
+					case ICPUImageView::ET_2D: return gli::TARGET_2D;
+					case ICPUImageView::ET_2D_ARRAY: return gli::TARGET_2D_ARRAY;
+					case ICPUImageView::ET_3D: return gli::TARGET_3D;
+					case ICPUImageView::ET_CUBE_MAP: return gli::TARGET_CUBE;
+					case ICPUImageView::ET_CUBE_MAP_ARRAY: return gli::TARGET_CUBE_ARRAY;
+					default: assert(0);
+				}
+			};
+
+			auto getFacesAndLayersAmount = [&]()
+			{
+				size_t layers, faces;
+				const auto arrayLayers = imageInfo.arrayLayers;
+
+				if (layersFlag)
+					layers = ((arrayLayers - 1) / 6) + 1;
+				else
+					layers = 1;
+				if (facesFlag)
+					faces = ((arrayLayers - 1) % 6) + 1;
+				else
+					faces = 1;
+
+				return std::make_pair(layers, faces);
+			};
+
+			auto gliFormatAndSwizzles = getTranslatedIRRFormat(imageViewInfo);
+			gli::target gliTarget = getTarget();
+			gli::extent3d gliExtent3d = {imageInfo.extent.width, imageInfo.extent.height, imageInfo.extent.depth};
+			size_t gliLevels = imageInfo.mipLevels;
+			std::pair<size_t, size_t> layersAndFacesAmount = getFacesAndLayersAmount();
+
+			gli::texture texture(gliTarget, gliFormatAndSwizzles.first, gliExtent3d, layersAndFacesAmount.first, layersAndFacesAmount.second, gliLevels, gli::texture::swizzles_type{ gliFormatAndSwizzles.second[0], gliFormatAndSwizzles.second[1], gliFormatAndSwizzles.second[2], gliFormatAndSwizzles.second[3] });
+
+			auto getFullSizeOfLayer = [&](const uint16_t mipLevel) -> uint64_t
+			{
+				auto region = image->getRegions().begin() + mipLevel;
+				const auto width = region->bufferRowLength == 0 ? region->imageExtent.width : region->bufferRowLength;
+				const auto height = region->bufferImageHeight == 0 ? region->imageExtent.height : region->bufferImageHeight;
+				const auto depth = region->imageExtent.depth;
+				return width * texelBlockByteSize * height * depth;
+			};
+
+			for (auto region = image->getRegions().begin(); region != image->getRegions().end(); ++region)
+			{
+				const auto ptrBeginningOfRegion = data + region->bufferOffset;
+				const auto layerSize = getFullSizeOfLayer(region->imageSubresource.mipLevel);
+
+				for (uint16_t layer = 0; layer < imageInfo.arrayLayers; ++layer)
+				{
+					const uint16_t gliLayer = layersFlag ? layer / 6 : 0;
+					const uint16_t gliFace = facesFlag ? layer % 6 : 0;
+					memcpy(texture.data(gliLayer, gliFace, region->imageSubresource.mipLevel), ptrBeginningOfRegion + (layer * layerSize), layerSize);
+				}	
+			}
+			
+			if (gli::save(texture, file->getFileName().c_str()))
```

Comment:
does it save ktx, dds, etc. properly?

Reply:
this won't work, you need to be able to write to virtual files... etc. so find a GLI function or provide one that overrides `fwrite` or whatever its using internally. Can't just go by filename and trust an external filesystem handler.

Maybe there's already something that uses an `ostream` or `istream` ?

Reply:
you're not saving to memory or loading from memory, you are passing the filename :man_facepalming:

Reply:
and passing a texture with all data needed to save an image
it performs saving to file

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/411#discussion_r367051458

Reply:
I want you to save to the `IWriteFile` only way to do it is
```cpp
void* someData = ALLOC(SIZE);
gli::save(texture,someData);
file->write(someData,SIZE);
FREE(someData);
```

likewise for reading

Reply:
well, in that case it's easy, but I'm forced to use `vector<char>` https://github.com/g-truc/gli/blob/master/gli/core/save_dds.inl#L126